### PR TITLE
[WIP] New package: spago-0.16.0

### PR DESCRIPTION
**Pull request by waynevanson**
I keep getting the error below and I cannot figure out how to solve it.

Anyone have any guidance? I've been trying for about 12 hours now on and off.
I've tried symlinks and depending on libtinfo files so far.

Error in log:

```sh
   SONAME: libtinfo.so.5 <-> UNKNOWN PKG PLEASE FIX!
```

Log:
```sh
[wayne@voidlinux void-packages]$ ./xbps-src pkg spago
=> xbps-src: updating repositories for host (x86_64)...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/x86_64-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/nonfree/x86_64-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/debug/x86_64-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/multilib/x86_64-repodata' ...
[*] Updating repository `https://alpha.de.repo.voidlinux.org/current/multilib/nonfree/x86_64-repodata' ...
=> xbps-src: updating software in / masterdir...
=> xbps-src: cleaning up / masterdir...
=> spago-0.16.0_1: removing autodeps, please wait...
=> spago-0.16.0_1: building for x86_64...
   [target] ncurses-libtinfo-libs-6.2_3: found (https://alpha.de.repo.voidlinux.org/current)
   [target] ncurses-libtinfo-devel-6.2_3: found (https://alpha.de.repo.voidlinux.org/current)
=> spago-0.16.0_1: installing target dependencies: ncurses-libtinfo-libs-6.2_3 ncurses-libtinfo-devel-6.2_3 ...
=> spago-0.16.0_1: skipping check (XBPS_CHECK_PKGS is disabled) ...
=> spago-0.16.0_1: running pre-pkg hook: 03-rewrite-python-shebang ...
=> spago-0.16.0_1: running pre-pkg hook: 04-generate-runtime-deps ...
   SONAME: libm.so.6 <-> glibc>=2.29_1
   SONAME: libstdc++.so.6 <-> libstdc++>=4.4.0_1
   SONAME: libz.so.1 <-> zlib>=1.2.3_1
   SONAME: libpthread.so.0 <-> glibc>=2.29_1
   SONAME: libtinfo.so.5 <-> UNKNOWN PKG PLEASE FIX!
   SONAME: librt.so.1 <-> glibc>=2.29_1
   SONAME: libutil.so.1 <-> glibc>=2.29_1
   SONAME: libdl.so.2 <-> glibc>=2.29_1
   SONAME: libgmp.so.10 <-> gmp>=5.0.1_1
   SONAME: libc.so.6 <-> glibc>=2.29_1
=> ERROR: spago-0.16.0_1: cannot guess required shlibs, aborting!
```

### CODE REVIEW

**Comment by sgn** [comment]
On 2020-09-06 18:01:53-0700, Wayne Van Son <notifications@github.com> wrote:
> I keep getting the error below and I cannot figure out how to solve it.
> 
> Anyone have any guidance? I&#39;ve been trying for about 12 hours now on and off.
> I&#39;ve tried symlinks and depending on libtinfo files so far.
> 
> Error in log:
> 
> ```sh
>    SONAME: libtinfo.so.5 &lt;-&gt; UNKNOWN PKG PLEASE FIX!
> ```

libtinfo is obsoleted, it was shipped with ncurses v5

Seems like this change fix it?
https://github.com/purescript/spago/commit/7dcd30d810d29c2bf2ae61c7947d183255b1844e

**Comment by waynevanson** [comment]
@sgn seems to me I should wait for them to issue a new build.

Are there any work arounds? I was trying to create a symlink but I didn't figure out which step and what commands to use, specifically for the template file.

**Comment by fosslinux** [comment]
Looks like you're in luck, the most recent commit:

https://github.com/purescript/spago/commit/7dcd30d810d29c2bf2ae61c7947d183255b1844e

Add https://github.com/purescript/spago/commit/7dcd30d810d29c2bf2ae61c7947d183255b1844e.patch as a patch.

File: srcpkgs/spago/template
Comment (fosslinux) [inline srcpkgs/spago/template:8]: unnessecary once fixed

File: srcpkgs/spago/template
Comment (fosslinux) [inline srcpkgs/spago/template:9]: `create_wrksrc` instead

**Review by fosslinux — CHANGES_REQUESTED** [review]
You should build from source instead of depending on their precompiled binary.

File: srcpkgs/spago/template
Comment (waynevanson) [reply srcpkgs/spago/template]: When building from binary, these are required so `stack` knows which GHC to build. It uses `libtinfo.so.6`

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:6]: This is not necessary if you handle the patch right.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:12]: Generally, `-devel` packages belong in `makedepends`.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:7]: This shouldn't be necessary with the patch you apply. Also, this would be the wrong package and the wrong place to declare it.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:29]: Rename `post_install` and remove the `vbin` line.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:22]: This function isn't necessary if the patch is handled properly and you use the right build style.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:25]: This patch should be placed in the `patches/` subdirectory of the package. `xbps-src` will apply the patch automatically after fetching.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:8]: This is not necessary if you use the right build style.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:11]: Are the binaries provided by `tar`, `xz` and `zlib` really necessary during build?

**Review by ahesford — CHANGES_REQUESTED** [review]
You should almost certainly be using `build_style=haskell-stack` to take care of properly setting up the `stack` environment for you.

File: srcpkgs/spago/template
Comment (ericonr) [reply srcpkgs/spago/template:11]: If the makefile or whatever calls `tar` for files with those compressions, it would make sense.

Comment (ahesford) [reply srcpkgs/spago/template:11]: I'm not familiar with `stack` builds and can't verify the need, but the author needs to be able to justify these build-time dependencies.

Comment (waynevanson) [reply srcpkgs/spago/template:11]: > Are the binaries provided by `tar`, `xz` and `zlib` really necessary during build?

Stack requires tar and xz and spago requires zlib as a dependency

**Comment by waynevanson** [comment]
This is now set up to build from source.

I've been having some trouble with the patches. I've never patched before. Any insight on why this might be happening? Check out TravisCI for some insight:

```sh
=> spago-0.16.0_1: verifying checksum for distfile '0.16.0.tar.gz'... OK.

=> spago-0.16.0_1: running do-extract hook: 00-distfiles ...

=> spago-0.16.0_1: extracting distfile(s), please wait...

=> spago-0.16.0_1: running do-patch hook: 00-patches ...

=> spago-0.16.0_1: patching: 7dcd30d810d29c2bf2ae61c7947d183255b1844e.patch.

=> ERROR: spago-0.16.0_1: do-patch_00-patches: 'patch -sl ${_args} -i ${_patch} 2> /dev/null' exited with 2

=> ERROR:   in _process_patch() at common/hooks/do-patch/00-patches.sh:34

=> ERROR:   in hook() at common/hooks/do-patch/00-patches.sh:43

=> ERROR:   in run_func() at common/xbps-src/shutils/common.sh:21

=> ERROR:   in run_pkg_hooks() at common/xbps-src/shutils/common.sh:242

=> ERROR:   in run_step() at common/xbps-src/shutils/common.sh:71

=> ERROR:   in main() at common/xbps-src/libexec/xbps-src-dopatch.sh:33

The command "if [ -z "$XLINT" ]; then docker exec -t void hostrepo/common/travis/build.sh $BOOTSTRAP $ARCH; fi" exited with 1.
```

**Comment by ericonr** [comment]
How did you generate these patches? You need to specify which file they are applied to. See patches in other packages.

If you use `git` to generate patches, it can be quite a bit easier.

**Comment by waynevanson** [comment]
@ericonr  The long name patch was copied from here: https://github.com/void-linux/void-packages/pull/24723#issuecomment-688129198

The second was made by myself. It was made by using `git diff changed.file original.file`

Is the error syntactical? We don't really get much info about why the error didn't go through

**Comment by ericonr** [comment]
You need the whole patch, but without the prefix in the path, so the patch from there:

```
From 7dcd30d810d29c2bf2ae61c7947d183255b1844e Mon Sep 17 00:00:00 2001
From: Fabrizio Ferrai <fabrizio.ferrai@gmail.com>
Date: Thu, 3 Sep 2020 12:09:30 +0200
Subject: [PATCH] Drop dependency on libtinfo (#684)

---
 stack.yaml | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/stack.yaml b/stack.yaml
index ae14f29f..d688acf5 100644
--- a/stack.yaml
+++ b/stack.yaml
@@ -44,3 +44,8 @@ extra-deps:
 allow-newer: true
 nix:
   packages: [zlib]
+
+# This is so we don't depend on libtinfo
+flags:
+  haskeline:
+    terminfo: false
\ No newline at end of file
```

Should become

```
From 7dcd30d810d29c2bf2ae61c7947d183255b1844e Mon Sep 17 00:00:00 2001
From: Fabrizio Ferrai <fabrizio.ferrai@gmail.com>
Date: Thu, 3 Sep 2020 12:09:30 +0200
Subject: [PATCH] Drop dependency on libtinfo (#684)

---
 stack.yaml | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/stack.yaml b/stack.yaml
index ae14f29f..d688acf5 100644
--- stack.yaml
+++ stack.yaml
@@ -44,3 +44,8 @@ extra-deps:
 allow-newer: true
 nix:
   packages: [zlib]
+
+# This is so we don't depend on libtinfo
+flags:
+  haskeline:
+    terminfo: false
\ No newline at end of file
```

**Review by ahesford — CHANGES_REQUESTED** [review]
First things first, please make sure to build the package locally before you push changes to your GitHub repository. GitHub CI is not intended as an exploratory tool; it provides a sanity check when you believe the package is ready for review. Obviously, if CI catches an unexpected mistake, you should push a fix and let it test again. However, when feeling your way through the development process, you'll get more meaningful messages locally, you won't consume limited testing resources that could be spent testing other PRs, and you minimize the flood of push notifications reviewers receive.

At a minimum, you should build a native package with `./xbps-src -f pkg spago` from the top level of your clone of `void-packages`. Because this is a new package and you are still exploring, you should try at least one cross build as well: `./xbps-src -f -a armv7l pkg spago`. Better still would also be to test a native and cross musl build, with
```
./xbps-src -m masterdir-musl binary-bootstrap x86_64-musl
./xbps-src -m masterdir-musl -f pkg spago
./xbps-src -m masterdir-musl -f -a aarch64-musl pkg spago
```
The first command should only be run once.

`xbps-src` does not recognize a `series` file. Packages are applied in lexicographical order, and your `series` file will be treated as a patch and fail because it is not valid. Also, you have two patches on `stack.yaml` that almost certainly conflict. Combine these two into a single patch and give it a more meaningful name than `44c44` or the hash of some commit in some other repository.

If there is a compelling reason to keep the patches separate (not really the case here), make sure that the second patch is diffed with respect to the file *as changed by the first* so it will apply sequentially.

**Comment by waynevanson** [comment]
@ahesford thank you for the explanation and patience, alongside my apologies for consuming CI resources.

To avoid this, I can add [ci skip] to the end of my commit message right?


I'll take onboard the advices given and try again tomorrow.

**Comment by ericonr** [comment]
> To avoid this, I can add [ci skip] to the end of my commit message right?

We do want final test results for this. The idea is that you will only push a new version into the PR if it builds correctly in your machine.

**Comment by Chocimier** [comment]
xbps-src do use `patches/series` file if present, as documented in Manual.md.

`git diff --no-prefix` can be used to create patch with paths expected by xbps-src.

**Comment by ahesford** [comment]
> xbps-src do use `patches/series` file if present, as documented in Manual.md.

Thanks for pointing this out; I overlooked that feature.

File: srcpkgs/spago/template
Comment (waynevanson) [reply srcpkgs/spago/template:8]: > We build Haskell package using stack from Stackage, generally the LTS versions. Haskell templates need to have host dependencies on `ghc` and `stack`, and set build style to haskell-stack.

from https://github.com/void-linux/void-packages/blob/master/Manual.md#haskell-packages

Do I still need to amend this change?

Comment (waynevanson) [reply srcpkgs/spago/template:8]: You're correct. It installs the dependencies.

**Comment by waynevanson** [comment]
As documented in the [optional variables section of the manual](https://github.com/void-linux/void-packages/blob/master/Manual.md#optional-variables), host dependencies cannot be set to a version. [This comment on GitHub](https://github.com/purescript/spago/issues/686#issuecomment-691197535) indicates that we need GHC to be of version `8.4` in order for it to build `spago`.

So far my working solution is to download the GHC specified by spago. This runs using `stack build` during phase `do_build()`.
Should downloading GHC via stack happen at a different phase? If so, when?

**Comment by waynevanson** [comment]
Should `x86_64-musl` and `i686` try to build on travis when `nocross="yes"`?
Do I need to disable these using the `archs` variable?

**Comment by fosslinux** [comment]
`nocross` indicates that in some way the package is not able to cross-compile. I believe you do want `archs` here; does spago only support x86_64?

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:5]: If you define your own `do_build` amd `do_install`, there is no value in using `build_style`, especially since the `haskell-stack` style pulls in the system GHC that you don't use. Just drop this line.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:7]: I thought the point of your `stack.yaml` patch was to drop the dependence on `ncurses-libtinfo-devel`.

File: srcpkgs/spago/template
Comment (ahesford) [inline srcpkgs/spago/template:15]: I think the problem here is that `stack` downloads a version of `ghc` that is only for `x86_64`. The template as is should either set `archs="x86_64"` or set `broken` when `"$XBPS_TARGET_MACHINE" != "x86_64"`. (The `archs` variable is really meant to indicate an upstream support limitation, not necessarily a consequence of XBPS packaging details or version incompatibilities.) It's not clear to me what option would be preferable here.

**Review by ahesford — COMMENTED** [review]
I offer some general advice here, but as I don't use Haskell and am not familiar with `stack` and `ghc`, I have no opinion about whether limiting the architecture by `archs` or `broken` and allowing this template to download its own version of `ghc` is acceptable. Hopefully somebody familiar with this ecosystem can offer an informed opinion.

File: srcpkgs/spago/template
Comment (ericonr) [reply srcpkgs/spago/template:15]: > I think the problem here is that stack downloads a version of ghc

Isn't this a problem on its own?

File: srcpkgs/spago/template
Comment (waynevanson) [reply srcpkgs/spago/template:7]: The patch is to ensure that `libtinfo.so.5` isn't used during the build process of spago.
There is a dependency called `haskeline` that requires it, which the patch addresses.

`libtinfo.so.6` is required so the correct version of `ghc` can be downloaded, which is what `ncurses-libtinfo-devel` installs via symlinks.

File: srcpkgs/spago/template
Comment (waynevanson) [reply srcpkgs/spago/template:15]: > > I think the problem here is that stack downloads a version of ghc
> 
> Isn't this a problem on its own?

Because ideally we want to use an xbps package, right? As I said in a comment, we need to download a specific, non-lts version  of GHC.

**Comment by fosslinux** [comment]
From the thread that you have linked, the problem seems to be that void packages GHC 8.8:

> It looks like you're on GHC 8.8, but Spago currently supports building only with GHC 8.4.

This is a relatively common issue in a variety of ecosystems and a thing distributions often have to solve. And, normally, the response from package maintainers is something along the lines of what you got:

> since maintaining a spago package with a newer GHC version than the one used in this project will be an ongoing source of pain for you because of breakages like this (we've seen this repeatedly with the compiler and its homebrew and nix packages).

I got something similar for Anki (we won't support rust stable just for distributions, use our official packages, etc, etc).

However, a lot of the time, someone has done the work to do it and it just isn't in a stable release yet, as is the case here.

This PR, https://github.com/purescript/spago/pull/647, was created for NixOS, and you can use the patch here. Add https://patch-diff.githubusercontent.com/raw/purescript/spago/pull/647.patch as a patch in XBPS (hot tip: adding .patch to the end of any commit or pull request URL on github will give you a unified diff file!)

Then, you can drop the janky different version of GHC and do it the proper way (using void's version of ghc).

**Comment by waynevanson** [comment]
I've pushed a version to you can see what I've done to try get this to work. The bulk of changes that make this ALMOST work are in the patches folder, with patches used from purescript/spago#647 mentioned by @fosslinux 

This is going to be used alongside purescript, so I'll add some comments about it here.
In Nix, purescript is not built form source, instead opting for use of binaries. I would like to use binaries for both `spago` and `purescript` for the time being. Would you accept this?

I think that getting this to build on `x86_64` is a good start, then gradually expand to other arches.

**Comment by fosslinux** [comment]
I don't think it would be accepted if it can be built from source. However I'm not a pkg commiter.

**Comment by waynevanson** [comment]
Here is the error I'm getting. It looks like it's in the Spago codebase and even comes with a suggested solution.

[14 of 29] Compiling Spago.GlobalCache

/builddir/spago-0.16.0/src/Spago/GlobalCache.hs:153:3: error:
    • Could not deduce (MonadThrow m)
        arising from a use of ‘assertDirectory’
      from the context: (MonadIO m, HasLogFunc env, MonadReader env m)
        bound by the type signature for:
                   getGlobalCacheDir :: forall (m :: * -> *) env.
                                        (MonadIO m, HasLogFunc env, MonadReader env m) =>
                                        m FilePath.FilePath
        at src/Spago/GlobalCache.hs:150:1-90
      Possible fix:
        add (MonadThrow m) to the context of
          the type signature for:
            getGlobalCacheDir :: forall (m :: * -> *) env.
                                 (MonadIO m, HasLogFunc env, MonadReader env m) =>
                                 m FilePath.FilePath
    • In a stmt of a 'do' block: _ <- assertDirectory globalCache
      In the expression:
        do globalCache <- liftIO
                            $ getXdgDirectory XdgCache "spago" <|> pure ".spago-global-cache"
           assertDirectory globalCache
           return globalCache
      In an equation for ‘getGlobalCacheDir’:
          getGlobalCacheDir
            = do globalCache <- liftIO
                                  $ getXdgDirectory XdgCache "spago" <|> pure ".spago-global-cache"
                 assertDirectory globalCache
                 return globalCache
    |
153 |   assertDirectory globalCache
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
Completed 180 action(s).

--  While building package spago-0.15.3 (scroll up to its section to see the error) using:
      /builddir/spago-0.16.0/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.0.1.0_ghc-8.8.4 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.0.1.0 build lib:spago exe:spago --ghc-options ""
    Process exited with code: ExitFailure 1

```sh
[14 of 29] Compiling Spago.GlobalCache

/builddir/spago-0.16.0/src/Spago/GlobalCache.hs:153:3: error:
    • Could not deduce (MonadThrow m)
        arising from a use of ‘assertDirectory’
      from the context: (MonadIO m, HasLogFunc env, MonadReader env m)
        bound by the type signature for:
                   getGlobalCacheDir :: forall (m :: * -> *) env.
                                        (MonadIO m, HasLogFunc env, MonadReader env m) =>
                                        m FilePath.FilePath
        at src/Spago/GlobalCache.hs:150:1-90
      Possible fix:
        add (MonadThrow m) to the context of
          the type signature for:
            getGlobalCacheDir :: forall (m :: * -> *) env.
                                 (MonadIO m, HasLogFunc env, MonadReader env m) =>
                                 m FilePath.FilePath
    • In a stmt of a 'do' block: _ <- assertDirectory globalCache
      In the expression:
        do globalCache <- liftIO
                            $ getXdgDirectory XdgCache "spago" <|> pure ".spago-global-cache"
           assertDirectory globalCache
           return globalCache
      In an equation for ‘getGlobalCacheDir’:
          getGlobalCacheDir
            = do globalCache <- liftIO
                                  $ getXdgDirectory XdgCache "spago" <|> pure ".spago-global-cache"
                 assertDirectory globalCache
                 return globalCache
    |
153 |   assertDirectory globalCache
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
Completed 180 action(s).

--  While building package spago-0.15.3 (scroll up to its section to see the error) using:
      /builddir/spago-0.16.0/.stack/setup-exe-cache/x86_64-linux/Cabal-simple_mPHDZzAJ_3.0.1.0_ghc-8.8.4 --builddir=.stack-work/dist/x86_64-linux/Cabal-3.0.1.0 build lib:spago exe:spago --ghc-options ""
    Process exited with code: ExitFailure 1
```

**Comment by waynevanson** [comment]
A dependency of Spago is Purescript, which blocks this PR until available.

**Comment by waynevanson** [comment]
Have to rename branch to remove `/` from branch name because it's causing conflicts in my builds.

Will update and send in at a later date, when `purs` is available.

**Pull request closed without merging**
